src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala

```diff
@@ -0,0 +1,423 @@
+      artifact.extract.foreach(uri.setExtract)
+      artifact.executable.foreach(uri.setExecutable)
+      artifact.destPath.foreach(uri.setOutputFile)
+
+      uri.build
+    }
+
+    commandInfo.addAllUris(uris.asJava)
+
+    val podEnvVars = podDefinition.env.collect{ case (k: String, v: EnvVarString) => k -> v.value }
+
+    val taskEnvVars = container.environment
+      .map(_.values)
+      .getOrElse(Map.empty[String, raml.EnvVarValueOrSecret])
+      .collect{ case (k: String, v: raml.EnvVarValue) => k -> v.value }
+
+    val hostEnvVar = Map("HOST" -> host)
+
+    val taskContextEnvVars = taskContextEnv(container, podDefinition.version, instanceId)
+
+    val labels = podDefinition.labels ++ container.labels.map(_.values).getOrElse(Map.empty[String, String])
+
+    val labelEnvVars = EnvironmentHelper.labelsToEnvVars(labels)
+
+    // Variables defined on task level should override ones defined at pod level.
+    // Therefore the order here is important. Values for existing keys will be overwritten in the order they are added.
+    val envVars = (podEnvVars ++
+      taskEnvVars ++
+      hostEnvVar ++
+      taskContextEnvVars ++
+      labelEnvVars ++
+      portsEnvVars)
+      .map {
+        case (name, value) =>
+          mesos.Environment.Variable.newBuilder.setName(name).setValue(value).build
+      }
+
+    commandInfo.setEnvironment(mesos.Environment.newBuilder.addAllVariables(envVars.asJava))
+
+    commandInfo
+  }
+
+  private[this] def computeContainerInfo(
+    podVolumes: Seq[raml.Volume],
+    container: raml.MesosContainer): mesos.ContainerInfo.Builder = {
+    val containerInfo = mesos.ContainerInfo.newBuilder
+      .setType(mesos.ContainerInfo.Type.MESOS)
+
+    container.volumeMounts.foreach { volumeMount =>
+      podVolumes.find(_.name == volumeMount.name).map { hostVolume =>
+        val volume = mesos.Volume.newBuilder
+          .setContainerPath(volumeMount.mountPath)
+
+        hostVolume.host.foreach(volume.setHostPath)
+
+        // Read-write mode will be used when the "readOnly" option isn't set.
+        val mode = if (volumeMount.readOnly.getOrElse(false)) mesos.Volume.Mode.RO else mesos.Volume.Mode.RW
+
+        volume.setMode(mode)
+
+        containerInfo.addVolumes(volume)
+      }
+    }
+
+    container.image.foreach { im =>
+      val image = mesos.Image.newBuilder
+
+      im.forcePull.foreach(forcePull => image.setCached(!forcePull))
+
+      im.kind match {
+        case raml.ImageType.Docker =>
+          val docker = mesos.Image.Docker.newBuilder.setName(im.id)
+
+          image.setType(mesos.Image.Type.DOCKER).setDocker(docker)
+        case raml.ImageType.Appc =>
+          // These labels are necessary for AppC images to work.
+          // Given that Docker only works under linux with 64bit,
+          // let's (for now) set these values to reflect that.
+          val labels = mesos.Labels.newBuilder
+            .addAllLabels(
+              Iterable(
```

Comment:
nit: can we extract this hard-coded sequence into a val? something like `val LinuxAmd64 = Seq(...)` ?